lib/MeshBVH.js

```diff
@@ -275,45 +47,34 @@ class MeshBVH extends MeshBVHNode {
 			node.boundingData = boundsToArray( bb );
 
 			// early out wif we've met our capacity
-			if ( tris.length <= options.maxLeafNodes ) {
+			if ( count <= options.maxLeafNodes ) {
 
-				node.tris = tris;
+				node.tris = ctx.tris;
+				node.offset = offset;
+				node.count = count;
 				return node;
 
 			}
 
 			// Find where to split the volume
-			const split = splitStrategy( bb, tris, centroids );
+			const split = ctx.getOptimalSplit( bb, offset, count, options.strategy );
 			if ( split.axis === - 1 ) {
 
-				node.tris = tris;
+				node.tris = ctx.tris;
+				node.offset = offset;
+				node.count = count;
 				return node;
 
 			}
 
-			// Collect the nodes for either side
-			const left = [];
-			const right = [];
-			for ( let i = 0; i < tris.length; i ++ ) {
-
-				const tri = tris[ i ];
-
-				if ( centroids[ tri * 3 + split.axis ] < split.pos ) {
-
-					left.push( tri );
-
-				} else {
-
-					right.push( tri );
-
-				}
-
-			}
+			const index = ctx.partition ( offset, count, split );
 
 			// create the two new child nodes
-			if ( ! left.length || ! right.length ) {
+			if ( index === offset || index === offset + count ) {
```

Comment:
This is basically the case where all of the triangles wind up on _one_ side of the partition, right? Shouldn't it be possible to know this ahead of time and avoid partitioning? Granted I'm not sure if there's a lot to gain from checking that.

Reply:
I agree, it feels like there is some duplication of concerns between this and the splitting code (which also can choose to return no-split-required.) Perhaps I will try to clean it up.

Reply:
On reflection, I think this is just a bad idea. Consider the AVERAGE and CENTER strategies. If it happens to be that all of the triangle centroids wind up on one side of the average of the longest edge -- which seems strange but possible given that the longest edge computation is using the full bounding box, not the centroid -- this code will terminate tree construction, even though maybe if you did a split again you would continue to make useful splits. So something should be different here.

Reply:
Would this be a problem in the AVERAGE split? In that case the average of all centroids are taken which would have to mean that some of the centroids are left on either side of the split. In the case of CENTER it would seem that this is just an artifact of choosing a less smart split strategy.

Ultimately I think this is okay.